### ISSUE TITLE
DOM elements get laid out in incorrect order

### ISSUE THREAD

**Issue by quentinadam** [preactjs/preact#3931]
- [x] Check if updating to the latest Preact version resolves the issue

**Describe the bug**

In some specific cases, the DOM elements get laid out in an incorrect order.

**To Reproduce**

I have tried to trim down my app to the bare minimum to show the issue appearing.
Here is a codesandbox: https://codesandbox.io/s/withered-bash-kcgn38?file=/App.tsx

From the code, you can read that the submit button should always be the last element on the page. When performing the following interactions, the submit button appears in between the other elements:

1. Please select a source country (for example France)
2. Please select a source city (for example Paris)
3. Please select a destination country (for example United States)
4. Please select a destination city (for example New York)
5. Now click back on the destination country as if you wanted to change it.
6. The submit button has now moved above the destination country combobox, which is incorrect.

Please see below screen recording:

https://user-images.githubusercontent.com/2349972/224035510-fafdbc30-454c-4826-9011-cf26b403024e.mp4

**Expected behavior**

The submit button should always be laid out as the last element on the page, like in the below (correct) sceen recording.
(the correct behaviour can be obtained by wrapping the "SubmitButtonLine" component in a "div")

https://user-images.githubusercontent.com/2349972/224035168-fcc93888-e09d-4a30-84ed-8c446d735fe0.mp4

**Comment by marvinhagemeister** [comment]
Looks like Preact gets confused when component functions are created inside the function of another component function. That's pretty much the worst case performance scenario for any virtual-dom based framework. Reason being is that the function is always a different one on every render. This means that the framework has to unmount the previous components of that and create it from scratch.

```tsx
function MyComponent() {
  // DON'T DO THIS.
  // Will be a new function on every render
  function OtherComponent() {
    return <p>this is bad</p>
  }

  // Because it's always a new function it won't be the same one that was used
  // for the last render. That means we have to destroy this component and re-create it.
  return <OtherComponent />
}
```

That said there is an argument to be made to have a better failure mode. Need to investigate more.

**Comment by marvinhagemeister** [comment]
FYI: Was able to make a smaller reproduction case based on the provided codesandbox.

```tsx
import { useState } from "preact/hooks";

export function App() {
  const [country, setCountry] = useState<string | undefined>(undefined);
  const [city, setCity] = useState<string | undefined>(undefined);

  function SourceConfirmationLine() {
    return <h1>Weird</h1>;
  }

  return (
    <div class="container">
      <SourceConfirmationLine />
      <ComboboxLine
        options={["click me"]}
        value={country}
        onSelect={(value) => {
          setCountry(value);
          setCity(undefined);
        }}
      />
      <ComboboxLine
        options={country ? ["Paris"] : undefined}
        value={city}
        onSelect={(value) => setCity(value)}
      />
      <Btn />
    </div>
  );
}

function Btn() {
  return (
    <b style="color: red;display:inline-block;padding-top: 1rem;">
      I should not move
    </b>
  );
}

export function ComboboxLine({
  options,
  value,
  onSelect,
}: {
  options: string[] | undefined;
  value: string | undefined;
  onSelect: (option: string | undefined) => void;
}) {
  if (options !== undefined) {
    return (
      <div class="form-line">
        {options.slice(0, 1).map((opt) => (
          <button
            class="combobox-item clickable"
            onClick={(e) => {
              e.preventDefault();
              onSelect(value ? undefined : opt);
            }}
          >
            {opt}
          </button>
        ))}
      </div>
    );
  }
  return null;
}
```

**Comment by JoviDeCroock** [comment]
Wondering about the worst case scenario here, currently we check whether the functional reference is equal to one in the children array. Should we have a fallback where we check whether we can find a function with the same name 😅

**Comment by JoviDeCroock** [comment]
This is fixed in 10.16.0
